help/flusspferd.1
.\" Copyright (c) 2008, 2009 Aristid Breitkreuz, Ash Berlin, Ruediger Sonderfeld
.\"
.\" Permission is hereby granted, free of charge, to any person obtaining a copy
.\" of this software and associated documentation files (the "Software"), to
.\" deal in the Software without restriction, including without limitation the 
.\" rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
.\" sell copies of the Software, and to permit persons to whom the Software is
.\" furnished to do so, subject to the following conditions:
.\"
.\" The above copyright notice and this permission notice shall be included in
.\" all copies or substantial portions of the Software.
.\"
.\" THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
.\" IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
.\" FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
.\" AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
.\" LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING 
.\" FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
.\" DEALINGS IN THE SOFTWARE.
.\"
.TH FLUSSPFERD 1 "July 2009"
.SH NAME
flusspferd \- interactive Javascript shell/repl
.SH SYNOPSIS
.B flusspferd\fR [\fIOPTION\fR] ... [\fIFILE\fR|-] [\fIARG\fR] ...
.SH DESCRIPTION
.PP
\fBflusspferd\fR is an interactive Javascript shell/repl based on the
Flusspferd Javascript library. It can be used to run and test Javascript code.
.SH OPTIONS
.TP
\fB-h\fR, \fB--help\fR
Displays a help message.
.TP
\fB-v\fR, \fB--version\fR
Print version and exit.
.TP
\fB-c\fR, \fB--config\fR \fIconfig-file\fR
Load config from file \fIconfig-file\fR.
.TP
\fB-i\fR, \fB--interactive\fR
Enter interactive mode (after all files are loaded).
.TP
\fB-0\fR, \fB--machine-mode\fR
(Interactive) Machine command mode (separator '\\0').
.TP
\fB-e\fR, \fB--expression\fR \fIexpression\fR
Evaluate the expression.
.TP
\fB-f\fR, \fB--file\fR \fIfile\fR
Run this file before standard script handling.
.TP
\fB-I\fR, \fB--include-path\fR \fIpath\fR
Add include path.
.TP
\fB-M\fR, \fB--module\fR \fImodule\fR
Load module.
.TP
\fB-m\fR, \fB--main\fR \fImodule\fR
Load module as the main module.
.TP
\fB--no-global-history\fR
Do not use a global history in interactive mode.
.TP
\fB--history-file\fR \fIfile\fR
Sets history file (default: ~/.flusspferd-history).
.TP
\fB--\fR
Stop processing options.
.SH EXAMPLES
.nf
  $ \fBflusspferd\fR
.fi
will run flusspferd in an interactive repl/shell mode.
.PP
.nf
  $ \fBflusspferd flusspferd.js\fR
.fi
.PP
will execute the file \fIflusspferd.js\fR and then quit.
.PP
.nf
  $ \fBflusspferd -i flusspferd.js\fR
.fi
will execute the file \fIflusspferd.js\fR and then run in an interactive
repl/shell mode.
.PP
.nf
  $ \fBflusspferd -e 'print("hello world");'\fR
.fi
will execute the expression 'print("hello world");'.
.SH SEE ALSO
For more information see our homepage at \fIhttp://www.flusspferd.org\fR
.SH AUTHOR
Flusspferd Team <\fIteam -AT- flusspferd -DOT- org\fR>


ChangeLog.txt
Flusspferd Change Log
=====================

Release 0.8
-----------

2009-09-10:
- Added bash_completion script for flusspferd

Release 0.7
-----------

2009-08-31:
- Use new getopt API for parsing interpreter / shell arguments.
- Add platform and xFlusspferdVersion constants to system module.

2009-07-30:
- Add (core) "getopt" plugin and C++ API.

2009-07-19:
- Remove environment plugin. Env variables are now only available via
  require('system').env (CommonJS compatability).

2009-07-04:
- Emacs mode is installed (again) and byte compiled if emacs is found.
- Line editors (libedit / readline / none) can now be selected via the cmake
  LINE_EDITOR option (also, similar to FORCE_PLUGINS, there is now
  FORCE_LINE_EDITOR).

2009-06-22:
- IO.stdin/stdout/stderr removed in favor of the same in the 'system' module.

2009-06-20:
- Added optional GNU/readline support (useful due to bugs in libedit).
- Added --no-global-history command line switch for deactivating the permanent
  history.
- Added --history-file command line switch for setting the history file.

2009-06-19:
- Add -I command line switch for specifying module paths.
- Add -M command line switch for loading modules (with 'require').
- Add permanent history to the shell.
- Add CommonJS 'system' module (https://wiki.mozilla.org/ServerJS/System).

2009-06-18:
- Remove Blob and use Binary everywhere.
- Adapted (method/class) names that contained Blob to use Binary instead.
- Eliminate the IO and XML global variables.

2009-06-12:
- Lower-case module names for included modules (it remains case-sensitive).
- FLUSSPFERD_LOADER gets passed the context object as its second parameter.
- Split FLUSSPFERD_LOADER into FLUSSPFERD_LOADER (2 parameters) and
  FLUSSPFERD_LOADER_SIMPLE (1 parameter).

Release 0.6.1
-------------

2009-06-13:
- Some plugins can now be enabled/disabled (PLUGIN_XXX) and required
  (FORCE_PLUGINS) in cmake.

Releases 0.5, 0.5.1, 0.6
------------------------

There was no change log for these releases.


programs/flusspferd.cpp
// vim:ts=2:sw=2:expandtab:autoindent:filetype=cpp:
/*
Copyright (c) 2008 Aristid Breitkreuz, Ruediger Sonderfeld

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.
*/

#include "flusspferd.hpp"
#include "flusspferd/spidermonkey/init.hpp"
#include "flusspferd/spidermonkey/object.hpp"
#include <boost/spirit/home/phoenix/core.hpp>
#include <boost/spirit/home/phoenix/bind.hpp>
#include <iostream>
#include <fstream>
#include <cstring>
#include <string>
#include <list>

#ifdef HAVE_EDITLINE
#include <editline/readline.h>
#ifdef HAVE_EDITLINE_HISTORY_H
#include <editline/history.h>
#endif
#endif

#include "config.h"

namespace phoenix = boost::phoenix;
namespace args = phoenix::arg_names;

class flusspferd_repl {
  bool interactive;
  bool machine_mode;
  std::istream in;

  bool config_loaded;
  char const *config_file;

  flusspferd::context co;
  flusspferd::current_context_scope scope;
  flusspferd::security &security;

  bool running;
  int exit_code;

  int argc;
  char ** argv;
  
  // Returns list of file(s) to execute
  std::list<std::string> parse_cmdline();
  void load_config();

  bool getline(std::string &source, const char* prompt = "> ");

  void quit(int code) {
    running = false;
    exit_code = code;
    throw flusspferd::js_quit();
  }

public:
  flusspferd_repl(int argc, char** argv);

  int run();
};

flusspferd_repl::flusspferd_repl(int argc, char **argv)
  : interactive(false),
    machine_mode(false),
    //file("typein"),
    in(std::cin.rdbuf()),
    config_loaded(false),
    // Default - can be changed by -c cmd line option
    config_file(INSTALL_PREFIX "/etc/flusspferd/jsrepl.js"),
    co(flusspferd::context::create()),
    scope(flusspferd::current_context_scope(co)),
    security(flusspferd::security::create(flusspferd::global())),
    running(false),
    exit_code(0),
    argc(argc),
    argv(argv)
{
  flusspferd::load_class<flusspferd::blob>();

  flusspferd::load_require_function();

  flusspferd::load_properties_functions();

  flusspferd::object g = flusspferd::global();
  flusspferd::create_native_function<void (int)>(
    g, "quit",
    phoenix::bind(&flusspferd_repl::quit, this, args::arg1));
  flusspferd::create_native_function(g, "gc", &flusspferd::gc);

  flusspferd::gc();
}

int flusspferd_repl::run() {
  std::list<std::string> files = parse_cmdline();

  if (!config_loaded)
    load_config();

  typedef std::list<std::string>::const_iterator iter;
  for (iter i = files.begin(), e = files.end(); i != e; ++i) {
    const std::string &file = *i;
    flusspferd::execute(file.c_str());
  }
  
  if (!interactive)
    return exit_code;

  std::string source;
  unsigned int line = 0;

  running = true;

  while (running && getline(source)) {
    unsigned int startline = ++line;

    for (;;) {
      JSBool compilable =
          JS_BufferIsCompilableUnit(
            flusspferd::Impl::current_context(),
            flusspferd::Impl::get_object(flusspferd::global()),
            source.data(),
            source.size());

      if (compilable)
        break;

      std::string appendix;
      getline(appendix, "? ");
      source += appendix;

      ++line;
    }

    try {
      flusspferd::value v = flusspferd::evaluate(source, "typein", startline);
      if (!v.is_undefined())
        std::cout << v << '\n';
    }
    catch(std::exception &e) {
      std::cerr << "ERROR: " << e.what() << '\n';
    }
    flusspferd::gc();
  }

  return exit_code;
}

void print_help(char const *argv0) {
  std::cerr << "usage: " << argv0 <<
    " [option] ... [file | -] [arg] ...\n\n"
    "Options\n"
    "    -h                       displays this message\n"
    "\n"
    "    -v\n"
    "    --version                print version and exit\n"
    "\n"
    "    -c <file>\n"
    "    --config <file>          load config from file\n"
    "\n"
    "    -i\n"
    "    --interactive            enter interactive mode (after files)\n"
    "\n"
    "    -0                       (interactive) machine command mode (separator '\\0')\n"
    "\n"
    "    -f <file>\n"
    "    --file <file>            run this file before standard script handling\n"
    "\n"
    "    --                       stop processing arguments\n\n";
}

void print_version() {
  std::cout << "flusspferd shell version: " << FLUSSPFERD_VERSION << '\n';
  std::cout << "flusspferd library version: " << flusspferd::version() << '\n';
  std::cout.flush();
}

void flusspferd_repl::load_config() {
  flusspferd::execute(config_file);
  config_loaded = true;

  // Get the prelude and execute it too
  flusspferd::value prelude = co.global().get_property("prelude");
  
  if (!prelude.is_undefined_or_null()) {
    flusspferd::execute(prelude.to_string().c_str());
  }
}

std::list<std::string> flusspferd_repl::parse_cmdline() {
  bool interactive_set = false;
  flusspferd::array args = flusspferd::create_array();
  co.global().set_property("arguments", args);

  std::list<std::string> files;

  int i=1;

  for (i = 1; i < argc; ++i) {
    if (argv[i][0] == '-') {
      if (std::strcmp(argv[i], "--") == 0)
      {
        i++;
        break;
      }
      else if (std::strcmp(argv[i], "-h") == 0 ||
          std::strcmp(argv[i], "--help") == 0)
      {
        print_help(argv[0]);
        throw flusspferd::js_quit();
      }
      else if (std::strcmp(argv[i], "-v") == 0 ||
          std::strcmp(argv[i], "--version") == 0)
      {
        print_version();
        throw flusspferd::js_quit();
      }
      else if (std::strcmp(argv[i], "-c") == 0 ||
               std::strcmp(argv[i], "--config") == 0)
      {
        ++i;
        if (i == argc) {
          print_help(argv[0]);
          std::string msg = "expected filename after ";
          msg += argv[i-1];
          msg += " option\n";
          throw std::runtime_error(msg);
        }
        config_file = argv[i];
      }
      else if (std::strcmp(argv[i], "-i") == 0 ||
               std::strcmp(argv[i], "--interactive") == 0)
      {
        interactive = true;
        interactive_set = true;
      }
      else if (std::strcmp(argv[i], "-0") == 0)
      {
        interactive = true;
        interactive_set = true;
        machine_mode = true;
      }
      else if (std::strcmp(argv[i], "-f") == 0 ||
               std::strcmp(argv[i], "--file") == 0)
      {
        ++i;
        if (i == argc) {
          print_help(argv[0]);
          std::string msg = "expected filename after ";
          msg += argv[i-1];
          msg += " option\n";
          throw std::runtime_error(msg);
        }
        std::string file = argv[i];
        if (!config_loaded)
          load_config();
        if (!interactive_set)
          interactive = false;
        files.push_back(file);
      }
    }
    else
      break; // Not an option, stop looking for one
  }

  if (i >= argc) {
    if (!interactive_set)
      interactive = files.empty();
  } else {
    // some cmd line args left.
    // first one is file
    // others go into arguments array
    std::string file = argv[i++];
    if (file == "-") {
      // TODO: Maybe check if stdin is actualy connected to a terminal?
      interactive = true;
    }
    else {
      files.push_back(file);
    }

    int x=0;
    for (; i < argc; ++i) {
      args.set_element(x++, flusspferd::string(argv[i]));
    }

  }

  return files;
}

bool flusspferd_repl::getline(std::string &source, const char* prompt) {
  if (machine_mode) {
    return std::getline(in, source, '\0');
  }
  else
#ifdef HAVE_EDITLINE
  if (interactive) {
    char* linep = readline(prompt);
    if (!linep) {
      std::cout << std::endl;
      return false;
    }
    if (linep[0] != '\0')
        add_history(linep);
    source = linep;
    source += '\n';
    return true;
  } 
  else
#endif
  {
    std::cout << prompt;
    return std::getline(in, source);
  }
}

int main(int argc, char **argv) {
  try {
    flusspferd::init::initialize();
    flusspferd_repl repl(argc, argv);
    return repl.run();
  } catch (flusspferd::js_quit&) {
  } catch (std::exception &e) {
    std::cerr << "ERROR: " << e.what() << '\n';
    return 1;
  }
}
